Mathlib/GroupTheory/Coxeter/Inversion.lean

```diff
@@ -0,0 +1,443 @@
+  rw [getD_rightInvSeq, eraseIdx_eq_take_drop_succ]
+  nth_rw 1 [← take_append_drop (j + 1) ω]
+  rw [take_succ]
+  rcases em (j < ω.length) with hj | nhj
+  · rw [get?_eq_get hj]
+    simp only [wordProd_append, wordProd_cons, mul_assoc]
+    simp
+  · rw [get?_eq_none.mpr (by linarith)]
+    simp
+
+theorem getD_leftInvSeq_mul_wordProd (ω : List B) (j : ℕ) :
+    ((lis ω).getD j 1) * π ω = π (ω.eraseIdx j) := by
+  rw [getD_leftInvSeq, eraseIdx_eq_take_drop_succ]
+  nth_rw 4 [← take_append_drop (j + 1) ω]
+  rw [take_succ]
+  rcases em (j < ω.length) with hj | nhj
+  · rw [get?_eq_get hj]
+    simp only [wordProd_append, wordProd_cons, mul_assoc]
+    simp
+  · rw [get?_eq_none.mpr (by linarith)]
+    simp
+
+theorem isRightInversion_of_mem_rightInvSeq {ω : List B} (hω : cs.IsReduced ω) {t : W}
+    (ht : t ∈ ris ω) : cs.IsRightInversion (π ω) t := by
+  constructor
+  · exact cs.isReflection_of_mem_rightInvSeq ω ht
+  · obtain ⟨⟨j, hj⟩, rfl⟩ := List.mem_iff_get.mp ht
+    rw [← List.getD_eq_get _ 1 hj, wordProd_mul_getD_rightInvSeq]
+    rw [cs.length_rightInvSeq] at hj
+    calc
+      ℓ (π (ω.eraseIdx j))
+      _ ≤ (ω.eraseIdx j).length   := cs.length_wordProd_le _
+      _ < ω.length                := by rw [← List.length_eraseIdx_add_one hj]; exact lt_add_one _
+      _ = ℓ (π ω)                 := hω.symm
+
+theorem isLeftInversion_of_mem_leftInvSeq {ω : List B} (hω : cs.IsReduced ω) {t : W}
+    (ht : t ∈ lis ω) : cs.IsLeftInversion (π ω) t := by
+  constructor
+  · exact cs.isReflection_of_mem_leftInvSeq ω ht
+  · obtain ⟨⟨j, hj⟩, rfl⟩ := List.mem_iff_get.mp ht
+    rw [← List.getD_eq_get _ 1 hj, getD_leftInvSeq_mul_wordProd]
+    rw [cs.length_leftInvSeq] at hj
+    calc
+      ℓ (π (ω.eraseIdx j))
+      _ ≤ (ω.eraseIdx j).length   := cs.length_wordProd_le _
+      _ < ω.length                := by rw [← List.length_eraseIdx_add_one hj]; exact lt_add_one _
+      _ = ℓ (π ω)                 := hω.symm
+
+theorem prod_rightInvSeq (ω : List B) : prod (ris ω) = (π ω)⁻¹ := by
+  induction' ω with i ω ih
+  · simp
+  · simp [rightInvSeq, ih, wordProd_cons]
+
+theorem prod_leftInvSeq (ω : List B) : prod (lis ω) = (π ω)⁻¹ := by
+  simp [leftInvSeq_eq_reverse_rightInvSeq_reverse, prod_reverse_noncomm]
+  have : List.map (fun x ↦ x⁻¹) (ris ω.reverse) = ris ω.reverse := calc
+    List.map (fun x ↦ x⁻¹) (ris ω.reverse)
+    _ = List.map id (ris ω.reverse)             := by
+        apply List.map_congr
+        intro t ht
+        exact (cs.isReflection_of_mem_rightInvSeq _ ht).inv
+    _ = ris ω.reverse                           := map_id _
+  rw [this]
+  nth_rw 2 [← reverse_reverse ω]
+  rw [wordProd_reverse]
+  exact cs.prod_rightInvSeq _
+
+theorem nodup_rightInvSeq_of_isReduced {ω : List B} (rω : cs.IsReduced ω) : List.Nodup (ris ω) := by
+  apply List.nodup_iff_get?_ne_get?.mpr
+  intro j j' j_lt_j' j'_lt_length dup
+  -- dup : get? (rightInvSeq cs ω) j = get? (rightInvSeq cs ω) j'
+  -- ⊢ False
```

Comment:
If you think these are useful, you might use `show` instead so lean checks them. Comments like this are likely to get out of date